blockchain/types/transaction.go

```diff
@@ -300,7 +300,6 @@ func (tx *Transaction) GasFeeCap() *big.Int {
 	return tx.data.GetPrice()
 }
 
-// This function is disabled because Kaia has no gas tip
 func (tx *Transaction) EffectiveGasTip(baseFee *big.Int) *big.Int {
```

Comment:
```suggestion
func (tx *Transaction) EffectiveGasTip(baseFee *big.Int) *big.Int {
	// effectiveGasPrice - baseFee = min(baseFee + tipCap, feeCap) - baseFee = min(tipCap, feeCap - baseFee)
	if tx.Type() == TxTypeEthereumDynamicFee {
		te := tx.GetTxInternalData().(TxInternalDataBaseFee)
		return math.BigMin(te.GetGasTipCap(), new(big.Int).Sub(te.GetGasFeeCap(), baseFee))
	}
	// effectiveGasPrice - baseFee = gasPrice - baseFee
	return new(big.Int).Sub(tx.GasPrice(), baseFee)
```

Reply:
Right, non-ethereumdynamicFee txtype also need to calculate tip. Btw, above doesn't consider baseFee==nil.  So I applied like below.

```
func (tx *Transaction) EffectiveGasTip(baseFee *big.Int) *big.Int {
	// effectiveGasPrice - baseFee = min(baseFee + tipCap, feeCap) - baseFee = min(tipCap, feeCap - baseFee)
	if baseFee != nil {
		// For EthereumDynamicFee TxType: min(GasTipCap, Sub(GasFeeCap,baseFee))
		// For Non-EthereumDynamicFee TxType: min(GasPrice, Sub(gasPrice, baseFee)
		tip := math.BigMax(big.NewInt(0), new(big.Int).Sub(tx.GasFeeCap(), baseFee))
		return math.BigMin(tx.GasTipCap(), tip)
	}

	return new(big.Int).Set(tx.GasTipCap())
}
```

Reply:
Check that tx.GasTipCap is correct for legacy tx types. Isn't it gasprice - basefee?

Reply:
For legacy tx types, tx.GasTipCap() or tx.GasFeeCap() returns tx.GasPrice. So EffectiveGasTip for legacy tx types return gasprice-basefee. Can I ask is this answer wanted?

Reply:
Understood. Thanks!